src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -554,30 +562,49 @@ public Task ProduceStartAndFireOnStarting()
 
         private void ProduceStart(bool appCompleted)
         {
-            if (_responseStarted) return;
-            _responseStarted = true;
+            if (HasResponseStarted)
+            {
+                return;
+            }
+
+            _requestProcessingStatus = RequestProcessingStatus.ResponseStarted;
 
             var statusBytes = ReasonPhrases.ToStatusBytes(StatusCode, ReasonPhrase);
 
             CreateResponseHeader(statusBytes, appCompleted);
         }
 
+        protected Task TryProduceEnd()
+        {
+            if (_requestProcessingStatus == RequestProcessingStatus.RequestStarted)
```

Comment:
Maybe call this TrySend400 and make this `if (_requestProcessingStatus == RequestProcessingStatus.RequestStarted && _requestRejected)`